src/Orleans/Threading/CancellationTokenManager.cs

```diff
@@ -0,0 +1,25 @@
+﻿using Orleans.Runtime;
+
+namespace Orleans.Threading
+{
+    internal class CancellationTokenManager
+    {
+        /// <summary>
+        /// Sets target activation to the found instances of type GrainCancellationToken
+        /// </summary>
+        /// <param name="arguments"> Grain method arguments list</param>
+        /// <param name="target"> Target grain reference</param>
+        public void SetGrainCancellationTokensTarget(object[] arguments, GrainReference target)
+        {
+            if (arguments == null) return;
+            for (var i = 0; i < arguments.Length; i++)
+            {
+                var argument = arguments[i];
+                if (argument is GrainCancellationToken)
+                {
+                    ((GrainCancellationToken) argument).TargetGrainReference = target;
+                }
+            }
```

Comment:
Since this code is in a hot path, re-write avoiding double cast and potentially using foreach, such as
```
foreach (var argument in arguments)
{
    var gct = argument as GrainCancellationToken;
    if (gct != null)
    {
        gct.TargetGrainReference = target;
    }
}
```